contrib/epee/include/net/levin_protocol_handler_async.h

```diff
@@ -190,115 +190,103 @@ class async_protocol_handler
-          epee::span<const uint8_t> fake;
-          cb(LEVIN_ERROR_CONNECTION_TIMEDOUT, fake, con.get_context_ref());
-          con.close();
-          con.finish_outer_call();
-        });
-        m_timer_started = true;
-      }
-    }
-    virtual ~anvoke_handler()
-    {}
-    callback_t m_cb;
-    async_protocol_handler& m_con;
+  protected:
     boost::asio::deadline_timer m_timer;
-    bool m_timer_started;
-    bool m_cancel_timer_called;
-    bool m_timer_cancelled;
-    uint64_t m_timeout;
-    int m_command;
-    virtual bool handle(int res, const epee::span<const uint8_t> buff, typename async_protocol_handler::connection_context& context)
+    async_protocol_handler& m_con;
+    const uint64_t m_timeout;
+    const int m_command;
+
+    virtual void do_handle(int res, const epee::span<const uint8_t> buff)=0;
+
+  public:
+    invoke_response_handler_base(uint64_t timeout, async_protocol_handler& con, int command)
+      : m_timer(con.m_pservice_endpoint->get_io_service()),
+        m_con(con),
+        m_timeout(timeout),
+        m_command(command)
+    {}
+
+    virtual ~invoke_response_handler_base() {}
+
+    //! Run user handler. Only call if `cancel_timer()` returns true.
+    void handle(int res, const epee::span<const uint8_t> buff)
     {
-      if(!cancel_timer())
-        return false;
-      m_cb(res, buff, context);
+      do_handle(res, buff);
       m_con.finish_outer_call();
-      return true;
     }
-    virtual bool is_timer_started() const
+
+    //! Cancel timer and run user handler if not previously invoked \return True if user handler invoked
+    bool cancel()
     {
-      return m_timer_started;
+      if (!cancel_timer())
+        return false;
+      handle(LEVIN_ERROR_CONNECTION_DESTROYED, nullptr);
+      return true;
     }
-    virtual void cancel()
+
+    //! Cancel timer but do not run user handler. \return True if user handler has not been invoked
+    bool cancel_timer()
     {
-      if(cancel_timer())
-      {
-        epee::span<const uint8_t> fake;
-        m_cb(LEVIN_ERROR_CONNECTION_DESTROYED, fake, m_con.get_context_ref());
-        m_con.finish_outer_call();
-      }
+      boost::system::error_code ignored{};
+      return m_timer.cancel(ignored);
     }
-    virtual bool cancel_timer()
+
+    // Attempt to adjust timeout for user handler execution. \return True if timer adjusted.
+    static bool reset_timer(const std::shared_ptr<invoke_response_handler_base>& self, const bool initial = false)
     {
-      if(!m_cancel_timer_called)
+      if (!self || (!initial && !self->cancel_timer()))
+        return false;
+      if (initial && !self->m_con.start_outer_call())
+        return false;
+
+      std::weak_ptr<invoke_response_handler_base> self_ = self;
+      self->m_timer.expires_from_now(boost::posix_time::milliseconds(self->m_timeout));
+      self->m_timer.async_wait([self_ = std::move(self_)](const boost::system::error_code& ec)
```

Comment:
I don't follow how `weak_ptr` is better here than the reference you had before. It doesn't impact the lifetime of the wrapped pointer.

Reply:
I'm not sure what you mean. The `weak_ptr` here adds little, which is why I didn't do it in the first attempt. It does guarantee cleanup of the callback before the `connection` destructor completes (instead of relying on `release_protocol()` to make that guarantee).

I could move the `weak_ptr` to the `deque` but that doesn't do much but add CPU cycles when invoking the callback.

Perhaps you were thinking of a change like #7345 where the callback has a `weak_ptr` to the `connection` object, instead of the reference. It might be possible to make that change with less changes, but I didn't see how.

Reply:
Oh and @jeffro256 recommended switching to a `weak_ptr` here, to be extra cautious against cycles (and see first paragraph from prior response).

Reply:
Sorry for the spam. After some thought - I shouldn't have made this change to `weak_ptr` here. I'll revert back to `shared_ptr` here. The `weak_ptr` here is always going to be valid because `start_outer_call()` forces a `shared_ptr` instance until `finish_outer_call()` is invoked.

The actual problem with lifetimes is fixed in #7345. This may have been what @jeffro256 was noticing, its somewhat messed up now because its not "correct" usage of `shared_ptr` / `weak_ptr` on the outer class.

This should be a `shared_ptr` and so should the one in the `deque<...>`. The internal reference in the callback here needs to be a `weak_ptr` to break the cycle and properly synchronize with destruction in all cases (it works now afaik, but is iffy and requires going through much more code to verify).

Reply:
@UkoeHB `weak_ptr` in this context wouldn't change the lifetime of the pointer, but it would allow detection if the lifetime ended too early to prevent segfaults. Also like @vtnerd mentioned, it would prevent `shared_ptr` cycles.